docs/shapes.md

```diff
@@ -137,3 +166,12 @@ index for each dimension. Linear indices are a single `int64` value which
 indexes into the buffer holding the array. See `shape_util.h` and
 `layout_util.h` in the same directory for utilities that simplify creation and
 manipulation of shapes and layouts.
+
+## Memory Space Identifiers
+
+In HLO, each array may be annotated with a memory space identifier, written as S(n).
```

Comment:
Also, looks like all of these are TPU-specific, but I'd like to confirm. If so, I can add a clear note about this.